src/NuGet.Services.AzureSearch/BatchPusher.cs

```diff
@@ -0,0 +1,257 @@
+                {
+                    var idAndValue = DequeueAndDecrement(actions, out int newCount);
+                    batch.Add(idAndValue.Value);
+
+                    if (newCount == 0)
+                    {
+                        allFinished.Add(NewIdAndValue(idAndValue.Id, _versionListDataResults[idAndValue.Id]));
+                        Guard.Assert(_versionListDataResults.Remove(idAndValue.Id), "The version list data result should have existed.");
+                    }
+                }
+
+                await IndexAsync(indexClient, batch);
+
+                if (allFinished.Any())
+                {
+                    var versionListIdSample = allFinished
+                       .OrderByDescending(x => x.Value.Result.VersionProperties.Count(v => v.Value.Listed))
+                       .Select(x => x.Id)
+                       .Take(5)
+                       .ToArray();
+                    _logger.LogInformation(
+                        "Updating {VersionListCount} version lists, including {IdSample}.",
+                        allFinished.Count,
+                        versionListIdSample);
+                    var stopwatch = Stopwatch.StartNew();
+
+                    foreach (var finished in allFinished)
+                    {
+                        _logger.LogDebug("Updating version list for package ID {PackageId}.", finished.Id);
+                        await _versionListDataClient.ReplaceAsync(
+                            finished.Id,
+                            finished.Value.Result,
+                            finished.Value.AccessCondition);
+                    }
+
+                    stopwatch.Stop();
+                    _logger.LogInformation(
+                        "Done updating {VersionListCount} version lists (took {Duration}).",
+                        allFinished.Count,
+                        stopwatch.Elapsed);
+                }
+            }
+
+            Guard.Assert(
+                !_versionListDataResults
+                    .Keys
+                    .Except(_idReferenceCount.Keys)
+                    .Any(),
+                "There are some version list data results without reference counts.");
+            Guard.Assert(
+                !_idReferenceCount
+                    .Keys
+                    .Except(_versionListDataResults.Keys)
+                    .Any(),
+                "There are some reference counts without version list data results.");
+        }
+
+        private async Task IndexAsync(
+            ISearchIndexClientWrapper indexClient,
+            IReadOnlyCollection<IndexAction<KeyedDocument>> batch)
+        {
+            if (batch.Count == 0)
+            {
+                return;
+            }
+
+            if (batch.Count > _options.Value.AzureSearchBatchSize)
+            {
+                throw new ArgumentException("The provided batch is too large.");
+            }
+
+            _logger.LogInformation(
```

Comment:
Could we add a metric that tracks the duration of the indexing operation, which index was updated, and whether the operaiton was successful? This would be useful for monitoring

Reply:
Tracked here: https://github.com/NuGet/NuGetGallery/issues/6664